orttraining/orttraining/core/optimizer/memory_alleviation.h

```diff
@@ -0,0 +1,285 @@
+
+/**
+@Class MemoryAlleviation
+
+Find recomputable subgraphs and enable according to user configs.
+*/
+
+class MemoryAlleviation : public GraphTransformer {
+ private:
+  using NodeOutputPort = std::pair<const Node*, int>;
+  using ActivationUsedMap = InlinedHashMap<std::string, std::pair<bool, bool>>;
+
+  /**
+   * @brief Level to control allowed operations during subgraph detecting.
+   * Level 0: only allow cheap-to-compute operations.
+   * Level 1: allow more expensive operations.
+   */
+  enum class ProbeLevel {
+    Basic = 0,
+    Advanced = 1,
+  };
+
+  /**
+   * @brief Type of memory reduction techniques.
+   */
+  enum class AlleviationType {
+    None = 0,  // Disabled.
+    Recompute = 1,
+  };
+
+  /**
+   * @brief Type of user config.
+   * type: type of memory reduction techniques.
+   * requested_count: the number of occurrences of a subgraph pattern for alleviation. -1 means apply all.
+   *   One example: if a subgraph pattern is found 3 times, and requested_count is set 2, then the 1st and 2nd subgraph in
+   *   typological order will be applied for alleviation. This is useful to avoid alleviating more memory than needed.
+   */
+  struct UserAlleviationConfig {
+    AlleviationType type;
+    int requested_count;
+  };
+
+  struct EntryOperatorConfig {
+    InlinedVector<int> input_arg_indices;  // input index to iterate further (bottom up)
+  };
+
+  /**
+   * @brief Struct to store properties of a specific subgraph.
+   */
+  struct AlleviationSubGraphDesc {
+    AlleviationSubGraphDesc() = default;
+
+    // A string to represent the subgraph, used as a unique "ID" for a unique subgraph.
+    std::string subgraph_representative_str;
+
+    InlinedHashMap<std::string, int> shape_str_frequency;  // shape string to frequency
+    UserAlleviationConfig user_alleviation_config;
+    int total_frequency{0};  // The occurrence of this subgraph pattern in the graph.
+
+    int applied_count{0};  // The number of times this subgraph pattern has been really applied in this transformer.
+    int skip_count{0};     // The number of times this subgraph instances will skipped in reversed topological order.
+    float saving_ratio{1.0f};
+  };
+
+  /**
+   * @brief A struct to maintain the information of recomputable subgraphs.
+   * Imagine we loop all nodes finding recomputable subgraphs, we want to store them first.
+   * Afterwards, we optionally pick up some of them to apply alleviation according to user configs.
+   *
+   * subgraph_descs is a map from subgraph string representation to its subgraph related configurations.
+   * recompute_graphs is a map from activation producer node pointers to its recomputable subgraph nodes.
+   *
+   * When we AddRecomputeSubGraphInstance, we must provider its corresponding subgraph desc in the parameter.
+   * Then we can know for each subgraph instance, what's the subgraph str representation, and what's the alleviation
+   * config.
+   */
+  struct AlleviationSubGraphStores {
```

Comment:
Can we rename this to RecomputeSubGraphs or RecomputeSubGraphStore?

Reply:
SubGraphStore

Reply:
Done.

Reply:
I found if someday we want to offload, we can still leverage this store. So remove the Recompute from all its names.